File: 167. Two Sum II - Input array is sorted.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    // hashtable: 19ms
    // vector<int> twoSum(vector<int>& numbers, int target) {
    //     unordered_map<int, int> add;    // store the complement of the member
    //     vector<int> ret;
    //     int len = numbers.size();
    //     bool found = false;
    //     for(int i = 0; i < len && (!found); i ++){
    //         int imp = target - numbers[i];
    //         unordered_map<int, int>::iterator it = add.begin();
    //         while(it != add.end()){
    //             if(it->first == numbers[i]){
    //                 ret.push_back(it->second);
    //                 ret.push_back(i + 1);
    //                 found = true;
    //                 break;
    //             }
    //             it ++;
    //         }
    //         add[imp] = i + 1;
    //     }
    //     return ret;
    // }

    // Two Pointers: 6ms
    vector<int> twoSum(vector<int>& numbers, int target){
        int low = 0, high = numbers.size() - 1;
        while(low < high){
            int sum = numbers[low] + numbers[high];
            if(sum == target) break;
            else if(sum < target) low += 1;
            else high -= 1;;
        }
        vector<int> ans(1, low + 1);
        ans.push_back(high + 1);
        return ans;
    }
};

int main(){
	// default test case
	int test[4] = {2, 7, 11, 15};
    vector<int> testCase(test, test+4);
	int target = 9;
    Solution sol;
	testCase = sol.twoSum(testCase, target);
    for(auto i : testCase)
        cout << i << "-";
	return 0;
}

File: 167. Two Sum II - Input array is sorted.py
class Solution(object):
    
    # brute force: O(N2) time, no space, TLE
    def twoSum(self, numbers, target):
        """
        :type numbers: List[int]
        :type target: int
        :rtype: List[int]
        """
        for i in xrange(len(numbers)):
            for j in xrange(i + 1, len(numbers)):
                if numbers[j] + numbers[i] == target:
                    return [i + 1, j + 1]
        return [-1, -1]
    
    # binary search: O(NlogN) time, O(1) space, TLE
    def twoSum(self, numbers, target):
        for i, num in enumerate(numbers):
            low, high = i + 1, len(numbers) - 1
            while low + 1 < high:
                mid = low + (high - low) / 2
                if numbers[mid] + num == target:
                    return [i + 1, mid + 1]
                elif numbers[mid] + num < target:
                    low = mid
                else:
                    high = mid
            if numbers[low] + num == target:
                return [i + 1, low + 1]
            if numbers[high] + num == target:
                return [i + 1, high + 1]
        return [-1, -1]
    
    # hash table: O(N) time & space
    def twoSum(self, numbers, target):
        index = {}
        for i, num in enumerate(numbers):
            if num in index:
                return [index[num], i + 1]
            else:
                index[target - num] = i + 1
        return [-1, -1]
                    
    # two pointers: O(N) time, O(1) space
    def twoSum(self, numbers, target):
        low, high = 0, len(numbers) - 1
        while low < high:
            currSum = numbers[low] + numbers[high]
            if currSum == target:
                return [low + 1, high + 1]
            elif currSum < target:
                low += 1
            else:
                high -= 1
        return [-1, -1]
                    